File: src/isalsr/core/labeled_dag.py
"""LabeledDAG -- Directed Acyclic Graph with node labels and cycle detection.

The central data structure of IsalSR. Extends IsalGraph's SparseGraph to:
1. Directed-only edges (expressions have data flow direction)
2. Node labels (NodeType: VAR, ADD, MUL, SIN, etc.)
3. Per-node metadata (var_index for VAR, const_value for CONST)
4. Dual adjacency lists (in + out) for efficient evaluation and cycle detection
5. Cycle detection on edge insertion (DFS reachability check)

Edge semantics: edge u -> v means "u provides input to v" (data flow direction).

Reference: IsalGraph's SparseGraph at
    /home/mpascual/research/code/IsalGraph/src/isalgraph/core/sparse_graph.py

Restriction: ZERO external dependencies. Only Python stdlib.
"""

from __future__ import annotations

import logging
from collections import deque

from isalsr.core.node_types import NodeType

log = logging.getLogger(__name__)


class LabeledDAG:
    """Directed acyclic graph with labeled nodes and cycle detection.

    Optimized for O(1) average edge insertion and membership testing.
    Always directed. Acyclicity enforced by ``add_edge``.

    Args:
        max_nodes: Upper bound on node count (pre-allocates storage).
    """

    __slots__ = (
        "_out_adj",
        "_in_adj",
        "_input_order",
        "_labels",
        "_node_data",
        "_node_count",
        "_edge_count",
        "_max_nodes",
    )

    def __init__(self, max_nodes: int) -> None:
        self._max_nodes: int = max_nodes
        self._out_adj: list[set[int]] = [set() for _ in range(max_nodes)]
        self._in_adj: list[set[int]] = [set() for _ in range(max_nodes)]
        # Ordered input lists: tracks the order in which in-edges were added.
        # Critical for non-commutative binary ops (SUB, DIV, POW) where
        # operand order determines semantics. V/v creates the first edge,
        # C/c creates subsequent edges.
        self._input_order: list[list[int]] = [[] for _ in range(max_nodes)]
        self._labels: list[NodeType | None] = [None] * max_nodes
        self._node_data: list[dict[str, int | float]] = [{} for _ in range(max_nodes)]
        self._node_count: int = 0
        self._edge_count: int = 0

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------

    @property
    def node_count(self) -> int:
        """Return the current number of nodes."""
        return self._node_count

    @property
    def edge_count(self) -> int:
        """Return the number of directed edges."""
        return self._edge_count

    @property
    def max_nodes(self) -> int:
        """Return the pre-allocated maximum node capacity."""
        return self._max_nodes

    def node_label(self, node: int) -> NodeType:
        """Return the NodeType label of *node*."""
        self._validate_node(node)
        label = self._labels[node]
        if label is None:
            raise ValueError(f"Node {node} has no label (should not happen)")
        return label

    def node_data(self, node: int) -> dict[str, int | float]:
        """Return per-node metadata dict (var_index, const_value, etc.)."""
        self._validate_node(node)
        return self._node_data[node]

    def set_const_value(self, node: int, value: float) -> None:
        """Set the const_value metadata for a node.

        Encapsulated setter for constant optimization. Avoids direct
        access to ``_node_data`` from external modules.

        Args:
            node: The node ID.
            value: The new constant value.
        """
        self._validate_node(node)
        self._node_data[node]["const_value"] = value

    def out_neighbors(self, node: int) -> frozenset[int]:
        """Return nodes that *node* has outgoing edges to (node -> target)."""
        self._validate_node(node)
        return frozenset(self._out_adj[node])

    def out_neighbors_raw(self, node: int) -> set[int]:
        """Return the underlying out-adjacency set for *node* (read-only use).

        Skips frozenset copy. Callers MUST NOT mutate the returned set.
        For performance-critical internal use (canonical backtracking).
        """
        return self._out_adj[node]

    def in_neighbors(self, node: int) -> frozenset[int]:
        """Return nodes that have edges into *node* (source -> node)."""
        self._validate_node(node)
        return frozenset(self._in_adj[node])

    def in_neighbors_raw(self, node: int) -> set[int]:
        """Return the underlying in-adjacency set for *node* (read-only use).

        Skips frozenset copy. Callers MUST NOT mutate the returned set.
        For performance-critical internal use (canonical backtracking).
        """
        return self._in_adj[node]

    def out_degree(self, node: int) -> int:
        """Return the number of outgoing edges from *node*."""
        self._validate_node(node)
        return len(self._out_adj[node])

    def in_degree(self, node: int) -> int:
        """Return the number of incoming edges to *node*."""
        self._validate_node(node)
        return len(self._in_adj[node])

    def has_edge(self, source: int, target: int) -> bool:
        """Return whether the directed edge source -> target exists."""
        self._validate_node(source)
        self._validate_node(target)
        return target in self._out_adj[source]

    def ordered_inputs(self, node: int) -> list[int]:
        """Return the ordered list of input source nodes for *node*.

        For non-commutative binary ops (SUB, DIV, POW), the order determines
        evaluation semantics: first element = first operand, second = second.
        The order is set by edge insertion order: V/v creates the first edge,
        C/c creates subsequent edges.

        For commutative ops (ADD, MUL) and unary ops, order is irrelevant
        but tracked for consistency.

        Args:
            node: The target node ID.

        Returns:
            List of source node IDs in insertion order.
        """
        self._validate_node(node)
        return list(self._input_order[node])

    def has_edge_unchecked(self, source: int, target: int) -> bool:
        """Return whether edge source -> target exists (no bounds check).

        For performance-critical internal use where node IDs are guaranteed valid.
        """
        return target in self._out_adj[source]

    def node_label_unchecked(self, node: int) -> NodeType:
        """Return node label without bounds check.

        For performance-critical internal use where node ID is guaranteed valid.
        """
        return self._labels[node]  # type: ignore[return-value]

    def node_data_unchecked(self, node: int) -> dict[str, int | float]:
        """Return per-node metadata without bounds check.

        For performance-critical internal use where node ID is guaranteed valid.
        """
        return self._node_data[node]

    # ------------------------------------------------------------------
    # Mutation
    # ------------------------------------------------------------------

    def add_node(
        self,
        label: NodeType,
        var_index: int | None = None,
        const_value: float | None = None,
    ) -> int:
        """Add a new node with the given label and return its integer ID.

        Args:
            label: The NodeType for this node.
            var_index: Variable index for VAR nodes (0-based).
            const_value: Constant value for CONST nodes.

        Raises:
            RuntimeError: If the graph has reached ``max_nodes``.
        """
        if self._node_count >= self._max_nodes:
            raise RuntimeError(f"Maximum number of nodes reached: {self._max_nodes}")
        node_id: int = self._node_count
        self._labels[node_id] = label
        data: dict[str, int | float] = {}
        if var_index is not None:
            data["var_index"] = var_index
        if const_value is not None:
            data["const_value"] = const_value
        self._node_data[node_id] = data
        self._node_count += 1
        return node_id

    def add_edge(self, source: int, target: int) -> bool:
        """Add directed edge source -> target if it preserves acyclicity.

        Edge semantics: source provides input to target (data flow).

        Returns:
            True if the edge was added. False if it would create a cycle
            or is a duplicate of an existing edge.

        Raises:
            IndexError: If either node ID is out of range.
        """
        self._validate_node(source)
        self._validate_node(target)

        # Duplicate check.
        if target in self._out_adj[source]:
            return False

        # Self-loop always creates a cycle in a DAG.
        if source == target:
            return False

        # Cycle check: would adding source -> target create a path target -> ... -> source?
        if self._has_path(target, source):
            return False

        self._out_adj[source].add(target)
        self._in_adj[target].add(source)
        self._input_order[target].append(source)
        self._edge_count += 1
        return True

    def add_edge_unchecked(self, source: int, target: int) -> None:
        """Add directed edge source -> target WITHOUT cycle or duplicate checks.

        Use ONLY when acyclicity is guaranteed by construction, e.g., when
        *target* is a freshly created node with no outgoing edges (V/v in
        canonical backtracking). Skips BFS reachability check for O(1)
        insertion instead of O(V+E).

        IMPORTANT: Still tracks _input_order for operand ordering (B9).

        Args:
            source: Source node ID.
            target: Target node ID.
        """
        self._out_adj[source].add(target)
        self._in_adj[target].add(source)
        self._input_order[target].append(source)
        self._edge_count += 1

    def remove_edge(self, source: int, target: int) -> bool:
        """Remove directed edge source -> target.

        Returns:
            True if the edge existed and was removed, False otherwise.
            Needed for backtracking in canonical search.
        """
        self._validate_node(source)
        self._validate_node(target)

        if target not in self._out_adj[source]:
            return False

        self._out_adj[source].discard(target)
        self._in_adj[target].discard(source)
        if source in self._input_order[target]:
            self._input_order[target].remove(source)
        self._edge_count -= 1
        return True

    def undo_node(self) -> None:
        """Remove the last-added node and all its incident edges.

        Used for backtracking in canonical search. Defensively removes
        all incident edges before deleting the node.
        """
        if self._node_count == 0:
            return

        node_id = self._node_count - 1

        # Remove all outgoing edges.
        for target in list(self._out_adj[node_id]):
            self._in_adj[target].discard(node_id)
            if node_id in self._input_order[target]:
                self._input_order[target].remove(node_id)
            self._edge_count -= 1
        self._out_adj[node_id].clear()

        # Remove all incoming edges.
        for source in list(self._in_adj[node_id]):
            self._out_adj[source].discard(node_id)
            self._edge_count -= 1
        self._in_adj[node_id].clear()

        self._labels[node_id] = None
        self._node_data[node_id] = {}
        self._input_order[node_id] = []
        self._node_count -= 1

    # ------------------------------------------------------------------
    # Cycle detection
    # ------------------------------------------------------------------

    def has_cycle_if_added(self, source: int, target: int) -> bool:
        """Check if adding edge source -> target would create a cycle.

        A cycle exists iff there is already a path from target to source
        in the current graph. Uses BFS from target.

        Complexity: O(V + E) per check.
        """
        if source == target:
            return True
        return self._has_path(target, source)

    def _has_path(self, start: int, end: int) -> bool:
        """Check if there is a directed path from *start* to *end* via BFS."""
        if start == end:
            return True

        visited: set[int] = {start}
        queue: deque[int] = deque([start])

        while queue:
            current = queue.popleft()
            for neighbor in self._out_adj[current]:
                if neighbor == end:
                    return True
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return False

    # ------------------------------------------------------------------
    # Graph algorithms
    # ------------------------------------------------------------------

    def topological_sort(self) -> list[int]:
        """Return nodes in topological order using Kahn's algorithm.

        Raises:
            ValueError: If the graph contains a cycle (should not happen
                        if add_edge is used correctly).
        """
        n = self._node_count
        in_deg = [len(self._in_adj[i]) for i in range(n)]
        queue: deque[int] = deque(i for i in range(n) if in_deg[i] == 0)
        result: list[int] = []

        while queue:
            node = queue.popleft()
            result.append(node)
            for neighbor in self._out_adj[node]:
                in_deg[neighbor] -= 1
                if in_deg[neighbor] == 0:
                    queue.append(neighbor)

        if len(result) != n:
            raise ValueError(
                f"Graph contains a cycle: topological sort produced {len(result)}/{n} nodes"
            )
        return result

    def output_node(self) -> int:
        """Return the unique non-VAR node with out_degree 0 (the expression root).

        For a valid expression DAG, this is the final output node whose value
        is the expression's result.

        Raises:
            ValueError: If there is no unique non-VAR sink node, or if
                there are multiple non-VAR sinks (ambiguous output).
        """
        sinks: list[int] = []
        for i in range(self._node_count):
            if len(self._out_adj[i]) == 0 and self._labels[i] != NodeType.VAR:
                sinks.append(i)

        if len(sinks) == 1:
            return sinks[0]
        if len(sinks) == 0:
            raise ValueError("No non-VAR sink nodes in graph")
        raise ValueError(
            f"Ambiguous output: {len(sinks)} non-VAR sink nodes "
            f"(IDs: {sinks}). A valid expression DAG must have "
            f"exactly one output node."
        )

    # ------------------------------------------------------------------
    # Isomorphism (label-aware backtracking)
    # ------------------------------------------------------------------

    def is_isomorphic(self, other: LabeledDAG) -> bool:
        """Test labeled DAG isomorphism with *other* via backtracking.

        Two labeled DAGs are isomorphic iff there exists a bijection between
        their node sets that preserves: (a) all directed edges, (b) all node
        labels, (c) operand order for non-commutative binary ops (SUB, DIV,
        POW). Variable nodes (VAR) are matched by their var_index, ensuring
        x_1 maps to x_1, x_2 maps to x_2, etc. (variables are distinguishable).

        Condition (c) is necessary because sin(x)-cos(x) and cos(x)-sin(x)
        have the same graph structure but different evaluation semantics.
        The canonical string encodes operand order (Bug Fix B9), so the
        isomorphism check must be consistent.
        """
        if not isinstance(other, LabeledDAG):
            return False

        # Normalize CONST creation edges before comparison so that DAGs
        # differing only in where CONST was created are considered isomorphic.
        self_dag = self.normalize_const_creation() if self._has_const_nodes() else self
        other_dag = other.normalize_const_creation() if other._has_const_nodes() else other

        if self_dag._node_count != other_dag._node_count:
            return False
        if self_dag._edge_count != other_dag._edge_count:
            return False

        n = self_dag._node_count
        if n == 0:
            return True

        # From here, use self_dag/other_dag (CONST-normalized if needed).
        sd, od = self_dag, other_dag

        # Check label multisets match (use .value for sorting since Enum lacks <).
        def _label_key(
            labels: list[NodeType | None], adj_out: list[set[int]], adj_in: list[set[int]], i: int
        ) -> tuple[str, int, int]:
            lbl = labels[i]
            return (lbl.value if lbl is not None else "", len(adj_out[i]), len(adj_in[i]))

        self_labels = sorted(_label_key(sd._labels, sd._out_adj, sd._in_adj, i) for i in range(n))
        other_labels = sorted(_label_key(od._labels, od._out_adj, od._in_adj, i) for i in range(n))
        if self_labels != other_labels:
            return False

        # Build initial mapping: VAR nodes are fixed (matched by var_index).
        mapping: dict[int, int] = {}
        used: set[int] = set()

        # Map VAR nodes by var_index (they are distinguishable).
        self_vars = {
            sd._node_data[i].get("var_index", -1): i
            for i in range(n)
            if sd._labels[i] == NodeType.VAR
        }
        other_vars = {
            od._node_data[i].get("var_index", -1): i
            for i in range(n)
            if od._labels[i] == NodeType.VAR
        }

        if self_vars.keys() != other_vars.keys():
            return False

        for var_idx in self_vars:
            u = self_vars[var_idx]
            v = other_vars[var_idx]
            mapping[u] = v
            used.add(v)

        # Order remaining nodes by (label, out_degree, in_degree) for pruning.
        remaining = [i for i in range(n) if i not in mapping]
        remaining.sort(
            key=lambda u: (
                str(sd._labels[u]),
                len(sd._out_adj[u]),
                len(sd._in_adj[u]),
            ),
            reverse=True,
        )

        other_remaining = [i for i in range(n) if i not in used]
        other_remaining.sort(
            key=lambda u: (
                str(od._labels[u]),
                len(od._out_adj[u]),
                len(od._in_adj[u]),
            ),
            reverse=True,
        )

        def _check_operand_order() -> bool:
            """Verify operand order for non-commutative binary ops."""
            from isalsr.core.node_types import BINARY_OPS

            for u, v in mapping.items():
                label = sd._labels[u]
                if label not in BINARY_OPS:
                    continue
                self_inputs = sd._input_order[u]
                other_inputs = od._input_order[v]
                if len(self_inputs) != len(other_inputs):
                    return False
                for si, oi in zip(self_inputs, other_inputs, strict=True):
                    if si in mapping and mapping[si] != oi:
                        return False
            return True

        def _backtrack(idx: int) -> bool:
            if idx == len(remaining):
                return _check_operand_order()
            u = remaining[idx]
            for v in other_remaining:
                if v in used:
                    continue
                # Labels must match.
                if sd._labels[u] != od._labels[v]:
                    continue
                # Degree must match.
                if len(sd._out_adj[u]) != len(od._out_adj[v]):
                    continue
                if len(sd._in_adj[u]) != len(od._in_adj[v]):
                    continue
                # Check edge consistency with already-mapped nodes.
                ok = True
                for u2, v2 in mapping.items():
                    # Check outgoing edges.
                    if (u2 in sd._out_adj[u]) != (v2 in od._out_adj[v]):
                        ok = False
                        break
                    # Check incoming edges.
                    if (u in sd._out_adj[u2]) != (v in od._out_adj[v2]):
                        ok = False
                        break
                if not ok:
                    continue

                mapping[u] = v
                used.add(v)
                if _backtrack(idx + 1):
                    return True
                del mapping[u]
                used.remove(v)
            return False

        return _backtrack(0)

    # ------------------------------------------------------------------
    # CONST creation edge normalization
    # ------------------------------------------------------------------

    def normalize_const_creation(self) -> LabeledDAG:
        """Return a new DAG with all CONST creation edges moved to x_1 (node 0).

        CONST nodes are evaluation-neutral leaves: they ignore in-edges and
        return ``const_value`` directly. But D2S requires every node to be
        reachable from a VAR via outgoing edges, so V/v creates a "creation
        edge" pointer → CONST. The choice of creation source is semantically
        irrelevant but produces different canonical strings.

        This normalization eliminates that redundancy by standardizing all
        CONST creation edges to come from node 0 (x_1). This is always valid
        because x_1 has no incoming edges (no cycle risk).

        The normalized DAG:
        - Computes the same function: eval(D) == eval(normalize(D))
        - Has deterministic CONST creation edges
        - Produces a unique canonical string for each equivalence class

        Returns:
            A new LabeledDAG with normalized CONST creation edges.
        """
        new = LabeledDAG(self._max_nodes)

        # Copy all nodes.
        const_nodes: set[int] = set()
        for i in range(self._node_count):
            label = self._labels[i]
            if label is None:
                continue
            data = self._node_data[i]
            new.add_node(
                label,
                var_index=int(data["var_index"]) if "var_index" in data else None,
                const_value=float(data["const_value"]) if "const_value" in data else None,
            )
            if label == NodeType.CONST:
                const_nodes.add(i)

        # Copy all edges EXCEPT creation edges (edges TO CONST nodes).
        for src in range(self._node_count):
            for tgt in self._out_adj[src]:
                if tgt in const_nodes:
                    continue  # Skip creation edges; will re-add from x_1.
                new.add_edge(src, tgt)

        # Add normalized creation edges: x_1 (node 0) -> each CONST.
        for c in sorted(const_nodes):
            new.add_edge(0, c)

        return new

    def _has_const_nodes(self) -> bool:
        """Return True if the DAG contains any CONST nodes."""
        return any(self._labels[i] == NodeType.CONST for i in range(self._node_count))

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def var_nodes(self) -> list[int]:
        """Return sorted list of VAR node IDs."""
        return [i for i in range(self._node_count) if self._labels[i] == NodeType.VAR]

    def non_var_nodes(self) -> list[int]:
        """Return sorted list of non-VAR node IDs."""
        return [i for i in range(self._node_count) if self._labels[i] != NodeType.VAR]

    def _validate_node(self, node: int) -> None:
        """Raise IndexError if *node* is out of range."""
        if node < 0 or node >= self._node_count:
            raise IndexError(f"Invalid node ID: {node} (node_count={self._node_count})")

    # ------------------------------------------------------------------
    # Dunder
    # ------------------------------------------------------------------

    def __repr__(self) -> str:
        label_counts: dict[str, int] = {}
        for i in range(self._node_count):
            lbl = self._labels[i]
            name = lbl.name if lbl is not None else "?"
            label_counts[name] = label_counts.get(name, 0) + 1
        labels_str = ", ".join(f"{k}={v}" for k, v in sorted(label_counts.items()))
        return (
            f"{self.__class__.__name__}("
            f"nodes={self._node_count}, "
            f"edges={self._edge_count}, "
            f"labels={{{labels_str}}})"
        )

File: code/docs/js/core/labeled-dag.js
/**
 * IsalSR — LabeledDAG Data Structure
 *
 * Directed acyclic graph with node labels, dual adjacency lists,
 * input-order tracking, and BFS-based cycle detection.
 * Port of isalsr.core.labeled_dag.LabeledDAG.
 */
(function () {
  'use strict';
  window.IsalSR = window.IsalSR || {};

  /**
   * @param {number} maxNodes - Upper bound on node count.
   */
  function LabeledDAG(maxNodes) {
    this._maxNodes = maxNodes;
    this._nodeCount = 0;
    this._labels = [];        // label per node (e.g. 'ADD', 'SIN', 'VAR')
    this._metadata = [];      // metadata per node (e.g. {varIndex: 0} or {constValue: 1.0})
    this._outAdj = [];        // out-neighbors (node provides input to these)
    this._inAdj = [];         // in-neighbors (these provide input to node)
    this._inputOrder = [];    // ordered input list per node (for binary ops)
    for (var i = 0; i < maxNodes; i++) {
      this._outAdj.push([]);
      this._inAdj.push([]);
      this._inputOrder.push([]);
      this._labels.push(null);
      this._metadata.push(null);
    }
  }

  LabeledDAG.prototype.nodeCount = function () { return this._nodeCount; };
  LabeledDAG.prototype.maxNodes = function () { return this._maxNodes; };

  LabeledDAG.prototype.getLabel = function (node) { return this._labels[node]; };
  LabeledDAG.prototype.getMetadata = function (node) { return this._metadata[node]; };
  LabeledDAG.prototype.getInputOrder = function (node) { return this._inputOrder[node]; };
  LabeledDAG.prototype.outNeighbors = function (node) { return this._outAdj[node]; };
  LabeledDAG.prototype.inNeighbors = function (node) { return this._inAdj[node]; };

  /**
   * Add a new node with the given label and metadata.
   * @returns {number} The new node ID.
   */
  LabeledDAG.prototype.addNode = function (label, metadata) {
    if (this._nodeCount >= this._maxNodes) {
      throw new Error('Maximum nodes reached: ' + this._maxNodes);
    }
    var id = this._nodeCount;
    this._labels[id] = label || null;
    this._metadata[id] = metadata || null;
    this._nodeCount++;
    return id;
  };

  /**
   * Add a directed edge source → target.
   * Does NOT check for cycles (caller must check with hasPath first).
   * @returns {boolean} true if edge was added, false if duplicate.
   */
  LabeledDAG.prototype.addEdge = function (source, target) {
    // Duplicate check
    if (this._outAdj[source].indexOf(target) !== -1) return false;
    this._outAdj[source].push(target);
    this._inAdj[target].push(source);
    this._inputOrder[target].push(source);
    return true;
  };

  /**
   * Check if a directed path exists from `from` to `to` using BFS.
   * Used for cycle detection: before adding edge u→v, check hasPath(v, u).
   */
  LabeledDAG.prototype.hasPath = function (from, to) {
    if (from === to) return true;
    var visited = {};
    var queue = [from];
    visited[from] = true;
    while (queue.length > 0) {
      var current = queue.shift();
      var neighbors = this._outAdj[current];
      for (var i = 0; i < neighbors.length; i++) {
        var next = neighbors[i];
        if (next === to) return true;
        if (!visited[next]) {
          visited[next] = true;
          queue.push(next);
        }
      }
    }
    return false;
  };

  /**
   * Get edge list for D3 rendering.
   * @returns {Array<{source: number, target: number}>}
   */
  LabeledDAG.prototype.getEdgeList = function () {
    var edges = [];
    for (var u = 0; u < this._nodeCount; u++) {
      for (var i = 0; i < this._outAdj[u].length; i++) {
        edges.push({ source: u, target: this._outAdj[u][i] });
      }
    }
    return edges;
  };

  /**
   * Convert to D3-friendly data.
   * @returns {{nodes: Array, edges: Array}}
   */
  LabeledDAG.prototype.toD3Data = function () {
    var nodes = [];
    for (var i = 0; i < this._nodeCount; i++) {
      var label = this._labels[i];
      var meta = this._metadata[i];
      var displayLabel = label || '?';
      if (label === 'VAR' && meta && meta.varIndex !== undefined) {
        displayLabel = 'x\u2081'.substring(0, 1) + String.fromCharCode(0x2081 + meta.varIndex);
      } else if (label === 'CONST') {
        displayLabel = 'k';
      }
      nodes.push({
        id: i,
        label: label,
        displayLabel: displayLabel,
        metadata: meta
      });
    }
    return { nodes: nodes, edges: this.getEdgeList() };
  };

  /**
   * Compute topological depth (distance from leaves/variables) for layout.
   * @returns {Array<number>} depth per node (0 = leaf/variable).
   */
  LabeledDAG.prototype.computeDepths = function () {
    var depths = [];
    var i;
    for (i = 0; i < this._nodeCount; i++) depths.push(-1);

    // BFS from nodes with no in-edges (roots of reversed DAG = leaves of original)
    var queue = [];
    for (i = 0; i < this._nodeCount; i++) {
      if (this._inAdj[i].length === 0) {
        depths[i] = 0;
        queue.push(i);
      }
    }

    while (queue.length > 0) {
      var node = queue.shift();
      var nextDepth = depths[node] + 1;
      var targets = this._outAdj[node];
      for (var j = 0; j < targets.length; j++) {
        var t = targets[j];
        if (nextDepth > depths[t]) {
          depths[t] = nextDepth;
          queue.push(t);
        }
      }
    }

    // Handle unreachable nodes
    for (i = 0; i < this._nodeCount; i++) {
      if (depths[i] === -1) depths[i] = 0;
    }

    return depths;
  };

  /**
   * Deep copy.
   */
  LabeledDAG.prototype.clone = function () {
    var copy = new LabeledDAG(this._maxNodes);
    copy._nodeCount = this._nodeCount;
    for (var i = 0; i < this._maxNodes; i++) {
      copy._labels[i] = this._labels[i];
      copy._metadata[i] = this._metadata[i] ? JSON.parse(JSON.stringify(this._metadata[i])) : null;
      copy._outAdj[i] = this._outAdj[i].slice();
      copy._inAdj[i] = this._inAdj[i].slice();
      copy._inputOrder[i] = this._inputOrder[i].slice();
    }
    return copy;
  };

  // Subscript digits for variable display
  var SUB_DIGITS = ['\u2081','\u2082','\u2083','\u2084','\u2085','\u2086','\u2087','\u2088','\u2089'];

  // Display names for expression rendering
  var EXPR_DISPLAY = {
    'ADD': '+', 'MUL': '\u00B7', 'SUB': '\u2212', 'DIV': '/',
    'POW': '^', 'SIN': 'sin', 'COS': 'cos', 'EXP': 'exp',
    'LOG': 'log', 'SQRT': '\u221A', 'ABS': 'abs', 'NEG': '\u2212',
    'INV': '1/', 'CONST': 'k'
  };

  var BINARY_OPS = { 'ADD': true, 'MUL': true, 'SUB': true, 'DIV': true, 'POW': true };
  var UNARY_OPS = { 'SIN': true, 'COS': true, 'EXP': true, 'LOG': true, 'SQRT': true, 'ABS': true, 'NEG': true, 'INV': true };

  /**
   * Convert this DAG to a human-readable mathematical expression string.
   * Finds root nodes (nodes with out-edges but no in-edges from other internal nodes,
   * or more precisely, nodes that are not inputs to any other node) and renders them.
   * If multiple roots exist, joins them with " ; ".
   * @returns {string} The mathematical expression.
   */
  LabeledDAG.prototype.toExpression = function () {
    if (this._nodeCount === 0) return '';

    var self = this;
    var visited = {};

    function renderNode(nodeId) {
      if (visited[nodeId]) return '(...)'; // cycle guard
      visited[nodeId] = true;

      var label = self._labels[nodeId];
      var meta = self._metadata[nodeId];

      // Leaf nodes
      if (label === 'VAR') {
        var idx = meta ? meta.varIndex : nodeId;
        visited[nodeId] = false;
        return 'x' + (SUB_DIGITS[idx] || String(idx + 1));
      }
      if (label === 'CONST') {
        visited[nodeId] = false;
        return 'k';
      }

      // Get inputs in insertion order
      var inputs = self._inputOrder[nodeId];
      if (!inputs || inputs.length === 0) {
        // No inputs yet — just show the operation name
        visited[nodeId] = false;
        return (EXPR_DISPLAY[label] || label) + '(?)';
      }

      var result;
      var sym = EXPR_DISPLAY[label] || label;

      if (UNARY_OPS[label]) {
        var child = renderNode(inputs[0]);
        if (label === 'NEG') {
          result = '(\u2212' + child + ')';
        } else if (label === 'INV') {
          result = '(1/' + child + ')';
        } else if (label === 'SQRT') {
          result = '\u221A(' + child + ')';
        } else {
          result = sym + '(' + child + ')';
        }
      } else if (BINARY_OPS[label]) {
        if (inputs.length === 1) {
          // Partially connected binary/variadic op
          result = sym + '(' + renderNode(inputs[0]) + ', ?)';
        } else {
          var parts = [];
          for (var i = 0; i < inputs.length; i++) {
            parts.push(renderNode(inputs[i]));
          }
          if (label === 'POW') {
            result = '(' + parts[0] + '^' + parts[1] + ')';
          } else {
            result = '(' + parts.join(' ' + sym + ' ') + ')';
          }
        }
      } else {
        // Unknown — render generically
        var args = [];
        for (var j = 0; j < inputs.length; j++) {
          args.push(renderNode(inputs[j]));
        }
        result = label + '(' + args.join(', ') + ')';
      }

      visited[nodeId] = false;
      return result;
    }

    // Find root nodes: nodes with out-neighbors that are NOT inputs to any other node
    // (i.e., nodes with no entries in any other node's _inputOrder)
    var isInput = {};
    for (var n = 0; n < this._nodeCount; n++) {
      for (var k = 0; k < this._inputOrder[n].length; k++) {
        isInput[this._inputOrder[n][k]] = true;
      }
    }

    // Roots = non-VAR, non-CONST nodes that nobody uses as input
    // If no such node, find the deepest nodes
    var roots = [];
    for (var r = 0; r < this._nodeCount; r++) {
      if (!isInput[r] && this._labels[r] !== 'VAR' && this._labels[r] !== 'CONST') {
        roots.push(r);
      }
    }

    // Fallback: if only VAR/CONST nodes, just list them
    if (roots.length === 0) {
      var vars = [];
      for (var v = 0; v < this._nodeCount; v++) {
        vars.push(renderNode(v));
      }
      return vars.join(', ');
    }

    var exprs = [];
    for (var e = 0; e < roots.length; e++) {
      exprs.push(renderNode(roots[e]));
    }
    return exprs.join(' ; ');
  };

  IsalSR.LabeledDAG = LabeledDAG;
})();
